### PR TITLE
Generation tutorial for Gemma model

### PR DESCRIPTION
**Pull request by pggPL**
# Description

I added the tutorials with finetuning and with generation for the Gemma model. Moreover I added few features that 
were neccessary to make my tutorials work.

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Changes

- Two new notebooks in the docs: one with finetuning for Gemma - analogous to the tutorial with Llama, one with generation for the Gemma,
- Generalized the kernel for rotary positional encoding to allow the sequences to start with different encoding positions,
- Added the kernel to effectively save key and value to the kv_cache,
- Expanded the class `InferenceParams` - which is responsible for caching k and v,
- Changed DotProductAttention to run THD attention when there are ragged tensors in kv_cache.

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

# Future work:

`TransformerLayer` does not support thd and it is a problem. The solutions right now works that way:
- one need to call `setup_before_new_input` before forward to indicate the sequence lengths,
- then one passes forward with TransformerLayer with `self_attn_format='thd'` and padded sequences with shape `bshd`,
- all layers get input is [b,s,*] format, not in [t, *] (including attention)
- `InferenceParams.retrieve_from_kv_cache` retrieves key_layer in bshd or ths format depending of `inference_params.qkv_format`,

As can be seen, it is quite messy workaround. How I think it should be done in the future:
- TransformerLayer supports thd and we do not need  `setup_before_new_input` at all,
- `InferenceParams` store lengths of cached sequences for each layer,
- for each TransformerLayer invocation, provided sequences are copied to the cache and lengths of cached sequences for this layer are updated,

To do this one will need to remove `save_to_kv_cache()` kernel and write `save_to_kv_cache_sbhd()` and  `save_to_kv_cache_thd()` (no bshd, because cache has shape sbhd both for bshd abd sbhd).
Logic of updating sequence lenghts needs to be moved from the `setup_before_new_input` into `save_to_kv_cache`.

It is worth noting that we need to take care of backwards compatibility. Right now generation works only for bsdh/sbdh and one needs to manually update `self.sequence_len_offset`. I think we can write setter which will update statistic for each of the layer when `sequence_len_offset` will be changed. 

If `TransformerLayer` support of thd will not be added in near future, I propose to write sequence lengths into `inference_params.cu_seqlens`, note that it is beta (in the future probably cu_seqlens will be added as an argument to the `TransformerLayer`). Then use `TransformerLayer` with bsdh. If `MultiHeadAttention` gets `inference_params.cu_seqlens != None`, it converts `bshd` with padding into `thd`, calls `save_to_kv_cache` etc. and run `DotProductAttention` with a `thd` and then converts output back to the `bshd`.

### CODE REVIEW

**Comment by sudhakarsingh27** [comment]
@cyanguwa, need your help in reviewing the `attention.py` and `test_fused_rope.py` files

File: transformer_engine/pytorch/transformer.py [hunk 1]
```diff
@@ -176,6 +176,10 @@ class TransformerLayer(torch.nn.Module):
                          head size. Note that these formats are very closely
                          related to the `qkv_format` in the `MultiHeadAttention`
                          and `DotProductAttention` modules.
+                         Notion: The experimental version of the 'thd' attention is supported
```
[the remark this answers is not in the document]
Comment (sudhakarsingh27) [reply transformer_engine/pytorch/transformer.py:179]: @cyanguwa, should we add `thd` as an allowable value for `attn_input_format` parameter? In this case, the actual input format is actually `bshd` but only the attention is done in `thd` format. What do you think?

File: transformer_engine/pytorch/csrc/extensions/generation.cu [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+template <typename scalar_t>
+__global__ void attention_copy_kernel(
+        scalar_t* cache_tensor,
+        int* seq_len,
+        int* incoming_seq_len,
+        scalar_t* hidden_tensor,
+        int max_incoming_seq_len,
+        int max_seq_len,
+        int b,
+        int s
```
Comment (pggPL) [inline transformer_engine/pytorch/csrc/extensions/generation.cu:19]: REname to hd or something

File: transformer_engine/pytorch/csrc/extensions/generation.cu [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    ) {
+    for (int batch_idx = blockIdx.x; batch_idx < b; batch_idx += gridDim.x) {
+        int to_copy = s * incoming_seq_len[batch_idx];
+        int offset = seq_len[batch_idx];
+
+        scalar_t* begin_cache_copy = cache_tensor + max_seq_len * s * batch_idx + s * offset;
+        scalar_t* begin_hidden_copy = hidden_tensor + s * batch_idx * max_incoming_seq_len;
```
Comment (pggPL) [inline transformer_engine/pytorch/csrc/extensions/generation.cu:26]: Either use cu_Seqlen or bring it up in the PR as future work.

File: transformer_engine/pytorch/csrc/extensions/generation.cu [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+        for (int i = threadIdx.x; i < to_copy; i += blockDim.x) {
+            *(begin_cache_copy + i) = *(begin_hidden_copy + i);
+        }
+    }
+}
+
+template <typename scalar_t>
+void attention_copy_launcher(torch::Tensor A, torch::Tensor seq_len, torch::Tensor incoming_seq_len,
+                    torch::Tensor B, int max_incoming_seq_len, int max_seq_len, int b, int s) {
+        attention_copy_kernel<<<16, 256, 0, at::cuda::getCurrentCUDAStream()>>>(
```
Comment (pggPL) [inline transformer_engine/pytorch/csrc/extensions/generation.cu:37]: griddim should be dynamic

File: tests/pytorch/fused_attn/test_fused_attn.py [hunk 5]
```diff
@@ -76,6 +76,12 @@ def _cudnn_version() -> Tuple[int, int, int]:
     return (major, minor, patch)
 
 
+def get_device_compute_capability() -> Tuple[int, int]:
```
Comment (cyanguwa) [inline tests/pytorch/fused_attn/test_fused_attn.py:79]: We have `from transformer_engine.pytorch.utils import get_device_compute_capability` at the top, so this function is redundant I think.

Comment (pggPL) [reply tests/pytorch/fused_attn/test_fused_attn.py]: idk why I pushed this file, I removed it, thanks

File: transformer_engine/pytorch/attention.py [hunk 6]
```diff
@@ -4710,21 +4964,44 @@ def forward(
...
+                    k_pos_emb = k_pos_emb[sequence_start:sequence_end, ...]
 
-                q_pos_emb = q_pos_emb[sequence_start:sequence_end, ...]
-                k_pos_emb = k_pos_emb[sequence_start:sequence_end, ...]
+                query_layer = apply_rotary_pos_emb(
+                    query_layer, q_pos_emb, self.qkv_format, fused=True)
+                key_layer = apply_rotary_pos_emb(
+                    key_layer, k_pos_emb, self.qkv_format, fused=True)
+        query_layer = query_layer.contiguous()
+        key_layer = key_layer.contiguous()
```
[the remark this answers is not in the document]
Comment (pggPL) [reply transformer_engine/pytorch/attention.py:5005]: no, I removed them

File: transformer_engine/pytorch/attention.py [hunk 7]
```diff
@@ -3633,6 +3893,9 @@ def forward(
                                first microbatch (since it is the first gradient being
                                produced)
         """
+        batch_size = key_layer.shape[0]
+        key_layer = key_layer.contiguous()
+        value_layer = value_layer.contiguous()
```
[the remark this answers is not in the document]
Comment (pggPL) [reply transformer_engine/pytorch/attention.py:3898]: apparenlty value is not contiguous, key_layer is. I moved call for value layers to one of the "ifs", which will affect the flow of the forward only for generation

[the remark this answers is not in the document]
Comment (pggPL) [reply transformer_engine/pytorch/attention.py:3898]: My kernel for attention copy does not work with non contiguous tensors. I think is not worth to change it, one need to write new separate kernels for THD and BSHD

File: transformer_engine/pytorch/attention.py [hunk 8]
```diff
@@ -5663,87 +5946,33 @@ def forward(
...
-                    attn_mask_type=attn_mask_type,
-                    window_size=window_size,
-                    alibi_slopes=alibi_slopes,
-                    cp_group=self.cp_group,
-                    cp_global_ranks=self.cp_global_ranks,
-                    cp_stream=self.cp_stream,
-                    max_seqlen_q=max_seqlen_q,
-                    max_seqlen_kv=max_seqlen_kv,
-                )
+        if self.attention_type == "self":
```
Comment (sudhakarsingh27) [inline transformer_engine/pytorch/attention.py:5949]: check if this line+5 are useful

**Comment by pluiez** [comment]
Hi @pggPL , any update on the gemma tutorial?

**Comment by pggPL** [comment]
@sudhakarsingh27 is working on the new version of the generation tutorial, which will be in the new PR.
